Add new Transform Select node

DolphinDream opened the pull request:
Behold the “Transform Select” node 🙌 😋 This sort of expands on the “mask vertices” node but it adds more to it to make it easier not only to separate/group the vertices but also to transform them (see the following demos). 

**Inputs**:

- **mask** # mask 0/1 with default repeat 0,1,0,1 if mask input is not linked
- **vertices**
- **polyEdges**
- **matrix T** # matrix to transform the True (mask=1) verts
- **matrix F** # matrix to transform the False (mask=0) verts

**Outputs**:

- **Vertices** # Original Vertices but transformed by MT and MF
- **PolyEdges** # original polyEdges (pass through)
- **PolyEdges O** # polyEdges in the original set connecting the T set with the F set
- **Vertices T** # vertices transformed by MT
- **PolyEdges T** # polyEdges in original set that are also in T set
- **Vertices F** # vertices transformed by MF
- **PolyEdges F** # polyEdges in original set that are also in F set

Note: The VEP selection does not work yet. This mode is to allow to mask either vertices, edges or polygons. 

![transofrmselectnode-ui](https://cloud.githubusercontent.com/assets/2083719/23840983/48f2d1d2-0780-11e7-8449-be197df0a35d.png)

DolphinDream commented:
The following demos show separating verts in a line into T and F sets (using a random mask) then transforms the verts in T and F sets. the magenta is the False set, green is the true set, and orange is the edges in the original edge set that bridge vertices between the two sets (the mirror is just for the kicks.. and for comparison to the unseparated line).

![transformselectnode-demo1](https://cloud.githubusercontent.com/assets/2083719/23826829/7408d6cc-0673-11e7-8a96-0542cf3fda8b.gif)

DolphinDream commented:
And a circle demo with scale transform:

![transformselectnode-demo2](https://cloud.githubusercontent.com/assets/2083719/23827045/21c7776a-0678-11e7-8593-3fb827d36420.gif)

DolphinDream commented:
And a mesh-select node + transform-select node combo :)

![transformselectnode-demo3](https://cloud.githubusercontent.com/assets/2083719/23827260/83f3ce98-067d-11e7-9a87-a2c43486d712.gif)

DolphinDream commented:
And a PlaneMK2 node + transform-select node combo :)

![transformselectionnode-demo4](https://cloud.githubusercontent.com/assets/2083719/23841042/c1c722fc-0780-11e7-92ba-eae047298977.gif)

enzyme69 commented:
I like your LOGIC, DolphinDream! Wow, I think I need to study this node.

jotero3d commented:
very interesting for reaction diffusion ;)

enzyme69 commented:
Would like to see Reaction Difussion as image and 3D points!

On Tue, 14 Mar 2017 at 4:46 pm, jotero3d <notifications@github.com> wrote:

> very interesting for reaction diffusion ;)
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/nortikin/sverchok/pull/1358#issuecomment-286328183>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADxQL5M-GzcauDwVvGR0y4O9ngrvW4QDks5rlim9gaJpZM4Ma001>
> .
>
-- 


------------------------------------------------------------
- Jimmy Gunawan
http://blendersushi.blogspot.com.au
http://mayaspiral.blogspot.com.au
http://houdoodles.blogspot.com.au
http://puppetar.blogspot.com.au

DolphinDream commented:
What is reaction diffusion ?

DolphinDream commented:
This kind of Reaction Diffusion ? 🤓 https://vimeo.com/176261480

DolphinDream commented:
Thinking more about the VEP potion to apply the mask to vertices, edges or polygons I can’t think of a really useful way I can do this. The trick is that the edges share verts and polygons share edges and verts and manipulating edges/polygons in True/False sets end up having to manipulate vertices belonging to both sets. Any ideas whether such masking would be still useful for E & P ?

DolphinDream commented:
If the matrix input socket has nothing connected what is the easiest way to have that matrix input be the identity matrix by default ?

portnov commented:
@DolphinDream 
just Matrix(). without parameters. Then you can convert it to list if you wish.

portnov commented:
I mean, mathutils.Matrix().

jotero3d commented:
DolphinDream @ yes that's it or here:
https://goo.gl/photos/u9hffPWbe2vTy5ti8

enzyme69 commented:
@jotero3d You got some awesome renders! @zeffii Maybe Jotero can add his artwork into Sverchok Gallery?
https://github.com/nortikin/sverchok/issues/1342

zeffii commented:
agreed, I welcome any similar renders by @jotero3d when sverchok is the main tool used.

jotero3d commented:
This is sverchok demo
https://goo.gl/photos/Hkv8qCmeyxbSco4U7

but this is my work, so not Blender sverchok
https://goo.gl/photos/u9hffPWbe2vTy5ti8

This is your work or? I will test it :)
http://code.activestate.com/recipes/579114-reaction-diffusion-simulation/

zeffii commented:
@jotero3d   https://gist.github.com/zeffii/9e156f0d37977fd1b0ca3c65d0ddc611   is only a slight modification of that.. but it's a little slow.

DolphinDream commented:
🙌 
```python
        identityMatrix = [[tuple(v) for v in Matrix()]]
```

DolphinDream commented:
@enzyme69 and this is what your “into the matrix 3” setup would look like with Transform Select node  😀 

![transformselect-intothematrix3](https://cloud.githubusercontent.com/assets/2083719/24076361/976395da-0c04-11e7-921e-23e916d87c3c.png)

enzyme69 commented:
@DolphinDream That's awesome!! This is why I need to make video tutorial and we need proper documentation overtime. But the development is high at this moment.

DolphinDream commented:
You’re right, @enzyme69 . Besides, this node does not do much vectorization for the time being. For instance if you want to replicate your house and have various transforms applied just to the selected vertex. I’ll have to think about how that would be possible with this node.

zeffii commented:
if you make the changes to the mixnumbers node a seperate PR i will merge that ASAP.

but I'm not in a hurry to add "transformSelectNode", maybe I don't appreciate what it does... ?

DolphinDream commented:
Whats a PR ?

zeffii commented:
Pull Request

DolphinDream commented:
Sorry @zeffii .. the mixnumbers checking was a mistake. It was supposed to be on a temporary branch. I rolled back the last commit. The rest is ready to go. 

Btw, did you mean you don't appreciate what the transformSelectNode does or what the mixNumbers additions to this branch does to the transformSelectNode ?

zeffii commented:
i was under the impression that the code changes in the MixNumbers node were bug fixes that you wanted to include in the main PR. (like an amendment)

zeffii commented:
the TSN node, seems like a solution to a very specific problem.. (one i've not had yet) and i think this is where my hesitation to merge comes from.

Also best practice is to not involve multiple nodes in a pull request, unless the PR is specifically about fixing a set of nodes. This PR was about TSN and somehow mixnumbers tagged along.  Maybe close the PR, remove changes to mixnum and pull again.

DolphinDream commented:
ok @zeffii .. here's the mixNumber node  PR https://github.com/nortikin/sverchok/pull/1440

DolphinDream commented:
@zeffii  I already said the mixnumber addition was a mistake commit to this branch.. i reverted back and added the mixnumber additions to a separate branch  and made a PR.

TNS is still under development at my end . I put it out there to see if anybody would find it interesting.

zeffii commented:
OK i didn't see that.

zeffii commented:
and by that i mean, Github still doesn't show (in this thread) that you removed  those commits from the PR , only now when i look at the "files changed" i see you've changed the content of the PR.  

weird.

zeffii commented:
Thanks for writing docs for it too.

DolphinDream commented:
Thanks for merging this @zeffii. We'll see of it's useful or not 🤓
